### DISCUSSION TITLE
Optimal image optimization settings for WordPress

### DISCUSSION THREAD

**Discussion by LukaszJaro** [ImageMagick/ImageMagick#3825] [Help]
Hi,

My host did install the imageMagick module, but upon uploading images to WordPress, or at least pngs, they actually grow triple in size compared to the original.

Is there some good default settings to use?

**Comment by snibgo** [comment]
Some samples would help us to understand your problem.

Are these photos of graphic images or what? What compression methods does WordPress use? Do you have control over those methods?

**Reply by LukaszJaro** [reply]
These are graphic pngs. WordPress uses imagemagik but falls back to gd library if the module is not installed. WP regenerates the images upon upload to different sizes and I do have control over the sizes and jpeg quality but no simple way for png quality. I'm going to open a ticket with my hosting provider to see if they can change the png quality setting, it sounds like the settings is at 100%?

**Comment by fmw42** [comment]
Were the PNGs 8-bit color or 24-bit color input? Perhaps they went from 8-bit color to 24-bit color?

**Reply by LukaszJaro** [reply]
Original image is 8 bit depth, I checked the regenerated version by WordPress and it's 32 bit, would this be partially the cause?

**Reply by fmw42** [reply]
Yes that would increase the size at least by 4.  Have WordPress save the pngs as "PNG00:filename.png" That should preserve the color format as 8-bit for 8-bit input images and 32-bit for 32-bit input images. See https://legacy.imagemagick.org/Usage/formats/#png_formats

**Reply by fmw42** [reply]
Please post the results of

`identify -verbose image.png`

on your input image or tell us if the 8-bit input image has transparency or not?

If on IM 7, make that 

`magick identify -verbose image.png`

**Reply by LukaszJaro** [reply]
Yes, there is transparency/alpha channel, is that also having impact?

**Reply by fmw42** [reply]
You may have to add -type palettematte (IM 6) or -type palettealpha (IM 7) along with PNG00:filename.png.  

What kind of processing are you doing between input and output in Wordpress or ImageMagick?

**Reply by LukaszJaro** [reply]
Thanks, it's processed using this class, on line 372 it sets some png defaults, below that it does something with the alpha channel and then on line 395 I noticed it's supposed to limit to 8bit?

https://github.com/xwp/wordpress-develop/blob/master/src/wp-includes/class-wp-image-editor-imagick.php#L372

Although through more research this appears to be an issue with large pngs https://core.trac.wordpress.org/ticket/48736

On line 249 "Do not scale (large) PNG images. May result in sub-sizes that have greater file size than the original"

 https://github.com/WordPress/WordPress/blob/master/wp-admin/includes/image.php#L249

Thanks for your help on this and quick response! I'm going to set the threshold to 1024 through a filter..the images I'm uploading are slightly below the 2560 default set by WP, but are still being impacted with bigger file sizes.

Filter here for reference https://developer.wordpress.org/reference/hooks/big_image_size_threshold/

This way I can avoid having images manipulated...and will open issue with WordPress community.

**Reply by fmw42** [reply]
Setting bit-depth to 8 in the way done in the script may not be adequate for PNG. I think the only way to ensure that is to preface the output name with PNG8:filename.png. But I am not sure if it will preserve you transparency if the input has transparency. But the best thing to resolve that would be simply to try it.

**Reply by fmw42** [reply]
Scaling the image, depending upon the scaling method, most likely will add more colors than the 256 limit for 8-bit color. So the image has to be forced to quantize colors to the 256 limit in order to save as 8-bits. Otherwise, it is saved as 24 or 32-bits. So specifying PNG8:filename.png will save as 8-bits.

**Reply by LukaszJaro** [reply]
Thank you, I opened a ticket with WordPress to see if they can implement something like this, I noticed when I remove alpha layer the images get saved as 24-bits instead of 32-bits.

**Reply by fmw42** [reply]
That would be correct, since alpha is an extra channel, RGBA so 4\*8-bits=32, but non-alpha is RGB so 3\*8-bits=24.

**Reply by snibgo** [reply]
@fmw42: In this forum, asterisk has a special meaning, and need to be escaped with backslash when we want a literal asterisk.

RGBA is 4 channels, and 4\*8= 32. RGB is 3 channels, and 3\*8=24.

@LukaszJaro: if you don't show a sample input to Worpress and the output it makes, all we can do is guess.

**Reply by fmw42** [reply]
@snibgo:  Thanks. I did not see that when I typed it. I had forgotten about the symbol meaning.

**Comment by snibgo** [comment]
@LukaszJaro: please show a sample input that you upload to Wordpress, and the output that Wordpress makes. If you can, also show the IM command that Wordpress uses.

**Reply by LukaszJaro** [reply]
Here's the image before upload to WordPress, I optimized it first by running it through tinypng which actually lowers the bit depth to 8, original non-optimized image is actually 32 bit:

![Working-at-Home-Vector-Isometric-Collection-00](https://user-images.githubusercontent.com/10998875/122047853-9d2e6880-cdae-11eb-8576-6c3eb4e7ef3f.png)

After upload to WordPress sample size:

![Working-at-Home-Vector-Isometric-Collection-00-1536x1485](https://user-images.githubusercontent.com/10998875/122048057-d1098e00-cdae-11eb-8c91-2a62bbfdb03d.png)

I opened a ticket with WordPress as well.

**Comment by fmw42** [comment]
I found that if you, say, blur the file and save with PNG8:filename.png, the 8-bit alpha format is preserved with only a small increase in file size. ImageMagick cannot achieve the same level of optimization as specialized programs such as you used. So the file size went from 105K to 112K. So have Wordpress change the code so that when 8-bits is desired, add the PNG8: prefix to the filename.

**Comment by snibgo** [comment]
As fmw42 says.

The image uploaded to Wordpress has 253 unique colours, so it can be stored with a palette, so each pixel needs one byte for storage. After the upload to Wordpress, it has 46523 colours, which is too much for a palette, so each byte needs 4 bytes for storage. This increase is caused by the resize, which generally increases the number of unique colours, even when the number of pixels reduces.

A possible workflow for Wordpress is: IF input is a paletted image, THEN use "PNG8:" for the output, to force a palette. However, this may reduce perceived quality.
